### PR TITLE
Partial schema validation

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
Experimental implementation of partial schema validation as described in https://github.com/cedar-policy/rfcs/pull/8.

The experimental implementation does not include the `additionalMemberOfTypes` and `additionalMemberOf` fields discussed in the RFC. I'm not convinced these are neccecary for the feature as a whole to be useful, and they require additional changes inside existing validation code.

I'm marking this ready for review as an experimental feature, so reviewers do not need to review the semantics of partial schema validation. They should be sure to review:

1. Changes which ensure partial schema validation is in fact experimental, i.e., it's properly placed behind the `partial-schema` feature.
2. That the code changes are should not effect the behavior of the strict validator.

### CODE REVIEW

**Comment by mwhicks1** [comment]
Some comments on the writeup:

> `RequestEnv` is now an enumeration which of either the request context as it was previously defined, or `RequestEnv::Unknown`.

This is a single catchall for handling missing actions, and also missing principals/resources associated with  a particular action. Would it be worthwhile to include a specific action in the cross-product, but associated with an _unknown_ principal or resource, in case the listed `principalTypes` or `resourceTypes` are incomplete?

> An undeclared entity type has an open attributes record, but a declared entity type without an explicitly defined attributes record has a closed attributes record by default. This is a bit weird. 

Why is it weird? Seems natural to me: The fact that you are explicitly declaring the entity in your schema, with an empty record, means you are taking a stand on it. If you want it to be open, you have to declare it as open, in the schema.

Same comment about not including a `memberOfTypes` element -- what you have seems right to me.

File: cedar-policy-validator/src/types.rs [hunk 1]
```diff
@@ -349,27 +424,46 @@ impl Type {
...
+            // UBs of ActionEntities are AnyEntity. So if we have an
+            // ActionEntity here its attrs are known.
             Type::EntityOrRecord(EntityRecordKind::ActionEntity { attrs, .. }) => {
                 attrs.iter().any(|(found_attr, _)| attr.eq(found_attr))
             }
-            // Other record or entity types always might have some attribute
-            // since the type may be the result of a least upper bound which
-            // could have dropped attributes from the type.
-            Type::EntityOrRecord(_) => true,
+            // A record will have an attribute if the attribute is in it's
```
Comment (mwhicks1) [inline cedar-policy-validator/src/types.rs:449]: it's --> its

File: cedar-policy-validator/src/types.rs [hunk 2]
```diff
@@ -349,27 +424,46 @@ impl Type {
...
+            // attributes map. Records computed as a LUB may have an open
+            // attributes record, but that is handled by the first match case.
+            Type::EntityOrRecord(EntityRecordKind::Record { attrs, .. }) => {
+                attrs.get_attr(attr).is_some()
+            }
+            _ => false,
```
Comment (mwhicks1) [inline cedar-policy-validator/src/types.rs:455]: What about the `Never` type?

File: cedar-policy-validator/src/types.rs [hunk 3]
```diff
@@ -904,9 +1022,40 @@ impl EntityRecordKind {
         }
     }
 
+    /// Return `true` if this entity or record may have additional undeclared
+    /// attributes.
+    pub(crate) fn has_open_attributes_record(&self, schema: &ValidatorSchema) -> bool {
+        match self {
```
Comment (mwhicks1) [inline cedar-policy-validator/src/types.rs:1028]: Redundant with code that seems to be doing the same thing above, i.e., `may_have_attr` ? Repeating comment from there: What about `Never` ?

File: cedar-policy-validator/src/types.rs [hunk 4]
```diff
@@ -904,9 +1022,40 @@ impl EntityRecordKind {
...
+                    // additional attributes.
+                    .unwrap_or(true)
+            }),
+        }
+    }
+
     pub(crate) fn get_attr(&self, schema: &ValidatorSchema, attr: &str) -> Option<AttributeType> {
         match self {
-            EntityRecordKind::Record { attrs } => attrs.get_attr(attr).cloned(),
+            EntityRecordKind::Record { attrs, .. } => attrs.get_attr(attr).cloned(),
```
Comment (mwhicks1) [inline cedar-policy-validator/src/types.rs:1058]: But if you have additional field types, shouldn't you possibly return `Never` for the attribute?

File: cedar-policy-validator/src/types.rs [hunk 5]
```diff
@@ -934,8 +1083,13 @@ impl EntityRecordKind {
...
         use EntityRecordKind::*;
         match (rk0, rk1) {
-            (Record { attrs: attrs0 }, Record { attrs: attrs1 }) => Some(Record {
+            (Record { attrs: attrs0, .. }, Record { attrs: attrs1, .. }) => Some(Record {
                 attrs: Attributes::least_upper_bound(schema, attrs0, attrs1),
+                // We could do an explicit check here to see if `attr0 == attr1`
+                // to keep `false` for `open_attributes`, but this result
+                // is already achieved due to the subtype check done by
+                // `Type::least_upper_bound`.
+                open_attributes: true,
```
Comment (mwhicks1) [inline cedar-policy-validator/src/types.rs:1092]: I'm not sure why you don't want to be precise. If I have an `if/then/else` and both branches return a record literal, I'd like to know that the record literal is closed. That way I'll catch `(if * then { f: 1 } else { f: 2 }).g` in partial schema validation as an error.

**Review by mwhicks1 — COMMENTED** [review]
I didn't review all the testing code, just to get this posted sooner. Will do that on the next review.

**Review by john-h-kastner-aws — COMMENTED** [review]
I'm making some tweaks to `in`, so I'll look at the comments there once I'm done with that

File: cedar-policy-validator/src/typecheck/test_utils.rs [hunk 6]
```diff
@@ -153,11 +153,27 @@ pub(crate) fn assert_policy_typechecks_for_mode(
...
         typechecker.mode = mode;
         let mut type_errors: HashSet<TypeError> = HashSet::new();
-        let typechecked = typechecker.typecheck_policy(&policy.into(), &mut type_errors);
+        let typechecked = typechecker.typecheck_policy(&policy, &mut type_errors);
         assert_eq!(type_errors, HashSet::new(), "Did not expect any errors.");
         assert!(typechecked, "Expected that policy would typecheck.");
+
+        // Ensure that partial schema validation doesn't cause any policy that
+        // should validate with a complete schema to no longer validate with the
+        // same complete schema.
```
Comment (khieta) [inline cedar-policy-validator/src/typecheck/test_utils.rs:165]: Useful property. We should set up a target for partial validation in `cedar-spec` that checks properties like this.

**Review by khieta — APPROVED** [review]
High-level comment: it would be helpful for review to split this PR into two: one for incidental refactoring and one for changes that need to be behind the partial validation flag.

But overall looks fine, so approving as-is.

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**